dart/dynamics/IkFast.cpp

```diff
@@ -0,0 +1,329 @@
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "dart/dynamics/IkFast.hpp"
+
+#include "dart/external/ikfast/ikfast.h"
+#include "dart/dynamics/BodyNode.hpp"
+#include "dart/dynamics/DegreeOfFreedom.hpp"
+
+namespace dart {
+namespace dynamics {
+
+namespace {
+
+//==============================================================================
+void convertTransform(
+    const IkReal* eetrans, const IkReal* eerot, Eigen::Isometry3d& tf)
+{
+  tf.linear()(0, 0) = eerot[0*3+0];
+  tf.linear()(0, 1) = eerot[0*3+1];
+  tf.linear()(0, 2) = eerot[0*3+2];
+  tf.linear()(1, 0) = eerot[1*3+0];
+  tf.linear()(1, 1) = eerot[1*3+1];
+  tf.linear()(1, 2) = eerot[1*3+2];
+  tf.linear()(2, 0) = eerot[2*3+0];
+  tf.linear()(2, 1) = eerot[2*3+1];
+  tf.linear()(2, 2) = eerot[2*3+2];
+
+  tf.translation() << eetrans[0], eetrans[1], eetrans[2];
+}
+
+//==============================================================================
+void convertTransform(
+    const Eigen::Isometry3d& tf,
+    std::array<IkReal, 3>& eetrans,
+    std::array<IkReal, 9>& eerot)
+{
+  eerot[0*3+0] = tf.linear()(0, 0);
+  eerot[0*3+1] = tf.linear()(0, 1);
+  eerot[0*3+2] = tf.linear()(0, 2);
+  eerot[1*3+0] = tf.linear()(1, 0);
+  eerot[1*3+1] = tf.linear()(1, 1);
+  eerot[1*3+2] = tf.linear()(1, 2);
+  eerot[2*3+0] = tf.linear()(2, 0);
+  eerot[2*3+1] = tf.linear()(2, 1);
+  eerot[2*3+2] = tf.linear()(2, 2);
+
+  eetrans[0] = tf.translation()[0];
+  eetrans[1] = tf.translation()[1];
+  eetrans[2] = tf.translation()[2];
+}
+
+//==============================================================================
+bool isFreeJoint(int numFreeParams, const int* freeParams, int index)
+{
+  for (auto i = 0; i < numFreeParams; ++i)
+  {
+    if (index == freeParams[i])
+      return true;
+  }
+
+  return false;
+}
+
+//==============================================================================
+void convertIkSolution(
+    const IkFast* ikfast,
+    int numJoints,
+    int numFreeParameters,
+    int* freeParameters,
+    const ikfast::IkSolutionBase<IkReal>& ikfastSolution,
+    InverseKinematics::Analytical::Solution& solution)
+{
+  const auto ik = ikfast->getIK();
+  const auto dofIndices = ikfast->getDofs();
+  const auto skel = ik->getNode()->getSkeleton();
+
+  std::vector<IkReal> solutionValues(numJoints);
+  std::vector<IkReal> freeValues(numFreeParameters);
+
+  ikfastSolution.GetSolution(solutionValues, freeValues);
+
+  bool limitViolated = false;
+
+  auto index = 0u;
+  solution.mConfig.resize(dofIndices.size());
+  for (auto i = 0u; i < solutionValues.size(); ++i)
+  {
+    if (isFreeJoint(numFreeParameters, freeParameters, i))
+      continue;
```

Comment:
You initialize `solution.mConfig` to an array of `dofIndices.size()` zeros. This `continue` skips incrementing `index`, which makes me think that the last `dofIndices.size() - numFreeParameters` elements of `solution.mConfig` will always be zero. This looks like a bug to me.

Also, I think it would be cleaner to implement this by converting `freeParameters` to an `std::vector<bool>` of flags that indicate whether each DOF is a free parameter or not.